README.md
# RSAA Crypt

Flutter plugin to use RSAA 1024 bit symmetric file and text on-device encryption in your app.

## Features

- 1024 bit secure symmetric encryption
- Supports all types of files for for File Encyption. 
- Supports Text and Base64 ciphers for Text Encryption.
- No change in encrypted files or Text size (in bytes using Text Mode).

### Installing

A step by step series of examples that tell you how to get a development env running

1. **Depend on it**

Add this to your package's pubspec.yaml file:

```
dependencies:
  rsaa_crypt: ^0.0.1
```

2. **Install it**

You can install packages from the command line with Flutter.

```
$ flutter pub get
```

3. **Import it**

Now in your Dart code, you can use:

```
import 'package:rsaa_crypt/rsaa_crypt.dart';
```

## Example 

```dart

import 'package:rsaa_crypt/rsaa_crypt.dart';

String key = '🐻🎌',
        message = 'Hello! Cześć! 🐻🎌 你好! ご挨拶！Привет! ℌ𝔢𝔩𝔩𝔬!',
        cipher,
        decipher;

// Text Encryption (Text Cipher)
RSAACrypt().encryptText(key, message).then((value) {
        cipher = Util().bytesToText(value);
      });
      
// Text Decryption (Text Cipher)
RSAACrypt().decryptText(key, Util().textToBytes(cipher)).then((value) {
        decipher = value;
      });
      
// Text Encryption (Base64 Cipher)
RSAACrypt().encryptText(key, message).then((value) {
        cipher = Util().bytesToBase64(value);
      });
      
// Text Decryption (Base64 Cipher)
RSAACrypt().decryptText(key, Util().base64ToBytes(cipher)).then((value) {
        decipher = value;
      });

// File Encryption
RSAACrypt().encryptFile(key, File(inputFilePath)).then((value) {
        Util().writeFile(encryptedFilePath, value);
      });

// File Decryption
RSAACrypt().decryptFile(key, File(encryptedFilePath)).then((value) {
        Util().writeFile(decryptedFilePath, value);
      });

```

## Contributing

Contributions will be acceptied at a later time.

## Authors

* **Vishal Robertson** - *Initial work* - [Robetron](https://github.com/Robetron)

See also the list of [contributors](https://github.com/Robetron/rsaa_crypt/contributors) who participated in this project.

## License

This project is licensed under the Apache License 2.0 - see the [LICENSE.md](LICENSE.md) file for details


CHANGELOG.md
## [0.0.2] - 1st March 2020.

* Bug Fixes


## [0.0.1] - 26th February 2020.

* Added code for Text and File Encryption and Decryption.


test/rsaa_crypt_test.dart
import 'dart:io';

import 'package:flutter_test/flutter_test.dart';
import 'package:logger/logger.dart';
import 'package:rsaa_crypt/rsaa_crypt.dart';

void main() {
  final Logger log = Logger();

  group('message-text', () {
    String key = '🐻🎌',
        message = 'Hello! Cześć! 🐻🎌 你好! ご挨拶！Привет! ℌ𝔢𝔩𝔩𝔬!',
        cipher,
        decipher;

    test('encrypt', () {
      RSAACrypt().encryptText(key, message).then((value) {
        cipher = Util().bytesToText(value);
      });
      log.d('Cipher Text: $cipher');
    });

    test('decrypt', () {
      RSAACrypt().decryptText(key, Util().textToBytes(cipher)).then((value) {
        decipher = value;
      });
      log.d('Deciphered Text: $decipher');
    });

    test('compare', () {
      expect(message, decipher);
    });
  });

  group('message-base64', () {
    String key = '🐻🎌',
        message = 'Hello! Cześć! 🐻🎌 你好! ご挨拶！Привет! ℌ𝔢𝔩𝔩𝔬!',
        cipher,
        decipher;

    test('encrypt', () {
      RSAACrypt().encryptText(key, message).then((value) {
        cipher = Util().bytesToBase64(value);
      });
      log.d('Cipher Text: $cipher');
    });

    test('decrypt', () {
      RSAACrypt().decryptText(key, Util().base64ToBytes(cipher)).then((value) {
        decipher = value;
      });
      log.d('Deciphered Text: $decipher');
    });

    test('compare', () {
      expect(message, decipher);
    });
  });

  group('file', () {
    String key = '🐻🎌',
        inputFilePath = 'assets/file-original.svg',
        encryptedFilePath = 'assets/file-encrypted.svg',
        decryptedFilePath = 'assets/file-decrypted.svg';

    test('encrypt', () {
      RSAACrypt().encryptFile(key, File(inputFilePath)).then((value) {
        Util().writeFile(encryptedFilePath, value);
      });
    });

    test('decrypt', () {
      RSAACrypt().decryptFile(key, File(encryptedFilePath)).then((value) {
        Util().writeFile(decryptedFilePath, value);
      });
    });

    test('compare', () {
      expect(File(inputFilePath).readAsBytesSync(),
          File(decryptedFilePath).readAsBytesSync());
    });
  });
}


lib/rsaa.dart
import 'dart:convert';

import 'package:basic_utils/basic_utils.dart';
import 'package:crypto/crypto.dart';
import 'package:logger/logger.dart';

import 'constants.dart';
import 'rsaa_crypt.dart';

class RSAA {
  static final Util util = Util();
  static final Logger log = Logger();
  List<List<int>> ruleBox = List.generate(NUM_RULES, (i) => List(NUM_RULES));

  String _hashKey(String inputKey) {
    return sha512.convert(utf8.encode(inputKey)).toString() +
        sha512.convert(utf8.encode(StringUtils.reverse(inputKey))).toString();
  }

  void ruleGen(String inputKey) {
    log.d('Rule Generation Started');

    List<List<num>> tempBox = List.generate(BOX_SIZE, (i) => List(BOX_SIZE));
    String key = _hashKey(inputKey);

    for (int i = 0, col = 0, row = 0; i < key.length; i++) {
      if (col == 16) {
        row++;
        col = 0;
      }
      tempBox[row][col++] = util.hexToInt(key[i]);
    }

    for (int line = 0; line < NUM_RULES; line++) {
      for (int pos = 0; pos < BOX_SIZE; pos++) {
        ruleBox[line][pos] = -1;
      }
    }

    for (int line = 0; line < NUM_RULES; line++) {
      int left = 0, right = 15;
      if (tempBox[line][left] == tempBox[line][right]) {
        ruleBox[line][tempBox[line][left]] = tempBox[line][left];
      } else {
        ruleBox[line][tempBox[line][right]] = tempBox[line][left];
        ruleBox[line][tempBox[line][left]] = tempBox[line][right];
      }
      while (left <= right) {
        while (left < 15 && ruleBox[line][tempBox[line][left]] != -1) {
          left++;
        }
        while (right > 0 && ruleBox[line][tempBox[line][right]] != -1) {
          right--;
        }
        if (tempBox[line][left] == tempBox[line][right]) {
          ruleBox[line][tempBox[line][left]] = tempBox[line][left];
        } else {
          ruleBox[line][tempBox[line][right]] = tempBox[line][left];
          ruleBox[line][tempBox[line][left]] = tempBox[line][right];
        }
      }
      for (int pos = 0; pos < BOX_SIZE; pos++) {
        if (ruleBox[line][pos] == -1) ruleBox[line][pos] = pos;
      }
    }

    log.d('Rule Generation Completed');
  }

  void encryptRSAA(List<String> hexInput) {
    int splitFactor = (hexInput.length < SPLIT_SETTING)
        ? 1
        : (hexInput.length ~/ SPLIT_SETTING);

    log.d('Encryption Started');

    // Iterations Loop
    for (var itr = 0; itr < ITERATIONS; itr++) {
      // Step 1: Cellular Automata Swap
      for (var index = 0; index < hexInput.length; index++) {
        var word1 = hexInput[index].substring(0, 1),
            word2 = hexInput[index].substring(1, 2);
        hexInput[index] =
            ruleBox[itr][util.hexToInt(word1)].toRadixString(HEX) +
                ruleBox[itr][util.hexToInt(word2)].toRadixString(HEX);
      }

      // Step 2: Left Shift by splitFactor
      List<String> tempInput = [];
      for (var i = 0; i < splitFactor; i++) {
        tempInput.add(hexInput[i]);
      }
      hexInput.removeRange(0, splitFactor);
      hexInput.addAll(tempInput);
    }

    log.d('Encryption Completed');
  }

  void decryptRSAA(List<String> hexInput) {
    int splitFactor = (hexInput.length < SPLIT_SETTING)
        ? 1
        : (hexInput.length ~/ SPLIT_SETTING);

    log.d('Decryption Started');

    // Iterations Loop
    for (var itr = ITERATIONS - 1; itr >= 0; itr--) {
      // Step 1: Right Shift by splitFactor
      List<String> tempInput = [];
      int startIndex = hexInput.length - splitFactor;
      for (var i = startIndex; i < hexInput.length; i++) {
        tempInput.add(hexInput[i]);
      }
      hexInput.removeRange(startIndex, hexInput.length);
      hexInput.insertAll(0, tempInput);

      // Step 1: Cellular Automata Swap
      for (var index = 0; index < hexInput.length; index++) {
        var word1 = hexInput[index].substring(0, 1),
            word2 = hexInput[index].substring(1, 2);
        hexInput[index] =
            ruleBox[itr][util.hexToInt(word1)].toRadixString(HEX) +
                ruleBox[itr][util.hexToInt(word2)].toRadixString(HEX);
      }
    }

    log.d('Decryption Completed');
  }
}


lib/util.dart
library rsaa_crypt;

import 'dart:io';
import 'dart:typed_data';

import 'package:logger/logger.dart';

class Util {
  int hexToInt(String hex) {
    switch (hex) {
      case 'a':
      case 'A':
      case '0a':
      case '0A':
        return 10;
      case 'b':
      case 'B':
      case '0b':
      case '0B':
        return 11;
      case 'c':
      case 'C':
      case '0c':
      case '0C':
        return 12;
      case 'd':
      case 'D':
      case '0d':
      case '0D':
        return 13;
      case 'e':
      case 'E':
      case '0e':
      case '0E':
        return 14;
      case 'f':
      case 'F':
      case '0f':
      case '0F':
        return 15;
      default:
        return int.parse(hex, radix: 10);
    }
  }

  Uint8List textToBytes(String sourceText) {
    var byteList = List<int>();
    sourceText.runes.forEach((rune) {
      if (rune >= 0x10000) {
        rune -= 0x10000;
        int firstWord = (rune >> 10) + 0xD800;
        byteList.add(firstWord >> 8);
        byteList.add(firstWord & 0xFF);
        int secondWord = (rune & 0x3FF) + 0xDC00;
        byteList.add(secondWord >> 8);
        byteList.add(secondWord & 0xFF);
      } else {
        byteList.add(rune >> 8);
        byteList.add(rune & 0xFF);
      }
    });
    return Uint8List.fromList(byteList);
  }

  String bytesToText(Uint8List sourceBytes) {
    StringBuffer stringBuffer = StringBuffer();
    for (int i = 0; i < sourceBytes.length;) {
      int firstWord = (sourceBytes[i] << 8) + sourceBytes[i + 1];
      if (0xD800 <= firstWord && firstWord <= 0xDBFF) {
        int secondWord = (sourceBytes[i + 2] << 8) + sourceBytes[i + 3];
        stringBuffer.writeCharCode(
            ((firstWord - 0xD800) << 10) + (secondWord - 0xDC00) + 0x10000);
        i += 4;
      } else {
        stringBuffer.writeCharCode(firstWord);
        i += 2;
      }
    }
    return stringBuffer.toString();
  }

  void getFileSize(Uint8List bytes) {
    Logger log = Logger();
    int fileSize = 0, fileLength = bytes.length;
    if (fileLength >= 1000000000) {
      fileSize = fileLength ~/ 1000000000;
      log.i('File Size: $fileSize GB');
    } else if (fileLength >= 1000000) {
      fileSize = fileLength ~/ 1000000;
      log.i('File Size: $fileSize MB');
    } else if (fileLength >= 1000) {
      fileSize = fileLength ~/ 1000;
      log.i('File Size: $fileSize KB');
    } else {
      log.i('File Size: $fileSize B');
    }
  }

  Uint8List readFile(File file) {
    return file.readAsBytesSync();
  }

  Future<File> writeFile(String outputPath, List<int> bytes) {
    return File(outputPath).writeAsBytes(bytes);
  }

  List<String> bytesToHexList(Uint8List bytes) {
    String temp;
    List<String> tempList = List<String>();
    for (var byte in bytes) {
      temp = byte.toRadixString(16);
      if (temp.length == 1)
        tempList.add("0" + temp);
      else
        tempList.add(temp);
    }
    return tempList;
  }

  Uint8List hexListToBytes(List<String> hexList) {
    List<int> tempList = List<int>();
    for (var hex in hexList) {
      tempList.add(int.parse(hex, radix: 16));
    }
    return Uint8List.fromList(tempList);
  }
}


lib/rsaa_crypt.dart
library rsaa_crypt;

import 'dart:convert';
import 'dart:io';
import 'dart:typed_data';

import 'package:logger/logger.dart';

import 'constants.dart';
import 'rsaa.dart';

class RSAACrypt {
  static final Util util = Util();
  static final RSAA rsaa = RSAA();
  List<List<int>> ruleBox = List.generate(NUM_RULES, (i) => List(NUM_RULES));

  Future<List<int>> encryptText(String key, String inputText) async {
    rsaa.ruleGen(key); // Generate rules
    Uint8List bytes = util.textToBytes(inputText); // Read Text as bytes
    List<String> hexInput =
        util.bytesToHexList(bytes); // Convert bytes to Hex List
    rsaa.encryptRSAA(hexInput); // Perform RSAA Decryption

    return util.hexListToBytes(hexInput); // Convert Hex to bytes and return
  }

  Future<String> decryptText(String key, Uint8List bytes) async {
    rsaa.ruleGen(key); // Generate rules
    List<String> hexInput =
        util.bytesToHexList(bytes); // Convert bytes to Hex List
    rsaa.decryptRSAA(hexInput); // Perform RSAA Decryption

    return util.bytesToText(
        util.hexListToBytes(hexInput)); // Convert Hex to bytes and return
  }

  Future<List<int>> encryptFile(String key, File inputFile) async {
    Uint8List bytes = util.readFile(inputFile); // Read File as bytes

    return await encryptBytes(key, bytes); // Convert Hex to bytes and return
  }

  Future<List<int>> decryptFile(String key, File inputFile) async {
    Uint8List bytes = util.readFile(inputFile); // Read File as bytes

    return await decryptBytes(key, bytes);
  }

  Future<List<int>> encryptBytes(String key, Uint8List bytes) async {
    rsaa.ruleGen(key); // Generate Cellular Automata Rules

    List<String> hexInput =
        util.bytesToHexList(bytes); // Convert bytes to Hex List
    rsaa.encryptRSAA(hexInput); // Perform RSAA Decryption

    return util.hexListToBytes(hexInput); // Convert Hex to bytes and return
  }

  Future<List<int>> decryptBytes(String key, Uint8List bytes) async {
    rsaa.ruleGen(key); // Generate Cellular Automata Rules

    List<String> hexInput =
        util.bytesToHexList(bytes); // Convert bytes to Hex List
    rsaa.decryptRSAA(hexInput); // Perform RSAA Decryption

    return util.hexListToBytes(hexInput); // Convert Hex to bytes and return
  }
}

class Util {
  int hexToInt(String hex) {
    switch (hex) {
      case 'a':
      case 'A':
      case '0a':
      case '0A':
        return 10;
      case 'b':
      case 'B':
      case '0b':
      case '0B':
        return 11;
      case 'c':
      case 'C':
      case '0c':
      case '0C':
        return 12;
      case 'd':
      case 'D':
      case '0d':
      case '0D':
        return 13;
      case 'e':
      case 'E':
      case '0e':
      case '0E':
        return 14;
      case 'f':
      case 'F':
      case '0f':
      case '0F':
        return 15;
      default:
        return int.parse(hex, radix: 10);
    }
  }

  Uint8List textToBytes(String sourceText) {
    var byteList = List<int>();
    sourceText.runes.forEach((rune) {
      if (rune >= 0x10000) {
        rune -= 0x10000;
        int firstWord = (rune >> 10) + 0xD800;
        byteList.add(firstWord >> 8);
        byteList.add(firstWord & 0xFF);
        int secondWord = (rune & 0x3FF) + 0xDC00;
        byteList.add(secondWord >> 8);
        byteList.add(secondWord & 0xFF);
      } else {
        byteList.add(rune >> 8);
        byteList.add(rune & 0xFF);
      }
    });
    return Uint8List.fromList(byteList);
  }

  String bytesToText(Uint8List sourceBytes) {
    StringBuffer stringBuffer = StringBuffer();
    for (int i = 0; i < sourceBytes.length;) {
      int firstWord = (sourceBytes[i] << 8) + sourceBytes[i + 1];
      if (0xD800 <= firstWord && firstWord <= 0xDBFF) {
        int secondWord = (sourceBytes[i + 2] << 8) + sourceBytes[i + 3];
        stringBuffer.writeCharCode(
            ((firstWord - 0xD800) << 10) + (secondWord - 0xDC00) + 0x10000);
        i += 4;
      } else {
        stringBuffer.writeCharCode(firstWord);
        i += 2;
      }
    }
    return stringBuffer.toString();
  }

  void getFileSize(Uint8List bytes) {
    Logger log = Logger();
    int fileSize = 0, fileLength = bytes.length;
    if (fileLength >= 1000000000) {
      fileSize = fileLength ~/ 1000000000;
      log.d('File Size: $fileSize GB');
    } else if (fileLength >= 1000000) {
      fileSize = fileLength ~/ 1000000;
      log.d('File Size: $fileSize MB');
    } else if (fileLength >= 1000) {
      fileSize = fileLength ~/ 1000;
      log.d('File Size: $fileSize KB');
    } else {
      log.d('File Size: $fileSize B');
    }
  }

  Uint8List readFile(File file) {
    return file.readAsBytesSync();
  }

  Future<File> writeFile(String outputPath, List<int> bytes) {
    return File(outputPath).writeAsBytes(bytes);
  }

  List<String> bytesToHexList(Uint8List bytes) {
    String temp;
    List<String> tempList = List<String>();
    for (var byte in bytes) {
      temp = byte.toRadixString(16);
      if (temp.length == 1)
        tempList.add("0" + temp);
      else
        tempList.add(temp);
    }
    return tempList;
  }

  Uint8List hexListToBytes(List<String> hexList) {
    List<int> tempList = List<int>();
    for (var hex in hexList) {
      tempList.add(int.parse(hex, radix: 16));
    }
    return Uint8List.fromList(tempList);
  }

  String bytesToBase64(List<int> bytes) {
    return Base64Codec().encode(bytes);
  }

  Uint8List base64ToBytes(String encoded) {
    return Base64Codec().decode(encoded);
  }
}
